File: Gold/1135 - 뉴스 전하기/BOJ1135.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <set>
#include <unordered_set>
#include <cstring>
using namespace std;

int n;
int a[51];
vector<int> tree[51];

int dfs(int cur) {
	vector<int> v;
	for (int next : tree[cur]) {
		v.push_back(dfs(next));
	}

	if (v.size() == 0) return 0;

	sort(v.rbegin(), v.rend());

	int ret = 0;
	for (int i = 0; i < v.size(); i++) {
		ret = max(ret, v[i] + i + 1);
	}

	return ret;
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> a[i];
		if (i == 0) continue;
		tree[a[i]].push_back(i);
	}

	cout << dfs(0);
}

File: Gold/1135 - 뉴스 전하기/BOJ1135.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class BOJ1135 {
    static int n;
    static ArrayList<Integer>[] tree;
    static int[] a;

    static int dfs(int cur){
        ArrayList<Integer> arr = new ArrayList<>();
        for(int next : tree[cur]){
            arr.add(dfs(next));
        }

        if(arr.size() == 0) return 0;

        arr.sort((a, b) -> {
            return b - a;
        });

        int ret = 0;
        for(int i = 0; i < arr.size(); i++){
            ret = Math.max(ret, arr.get(i) + i + 1);
        }

        return ret;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        tree = new ArrayList[n];
        a = new int[n];
        for(int i = 0; i < n; i++){
            tree[i] = new ArrayList<>();
        }

        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; i++){
             int parent = Integer.parseInt(st.nextToken());
            if(i == 0) continue;
            tree[parent].add(i);
        }

        System.out.println(dfs(0));
    }
}
